### ISSUE TITLE
OAMP baseline implementation problem

### ISSUE THREAD

**Issue by jiadong5** [STARainZ/CG-OAMP-NET#2]
Hi Xingyu,
My concern with your implementation is that you did not do the orthogonalization step in OAMP implementation, which is an important step for the orthogonality btw the NLE error and LE error.
I have also reviewed the Paper code of "H. He, C. Wen, S. Jin, and G. Y. Li, “Model-driven deep learning for MIMO detection,” IEEE Trans. Signal Process., vol. 68, pp. 1702–1715, Feb. 2020." Finding out that there is also no worth step in the NLE, could you please explain why you guys just abandoned the orth step?
Moreover, I have reviewed your network code, there isn't a neural network explicitly defined in the provided code. The code is just a huge optimization, could you please tell me why it's related to a neural network?
Best wishes,
Jade Jiadong Hong

**Comment by STARainZ** [comment]
Hi Jade,

Thanks for your comments. For the first concern, we agree that the orthogonalization step is critical in OAMP, and we allow this option in our code, cf the following line:

https://github.com/STARainZ/CG-OAMP-NET/blob/44b5f92b7428b2bd8939ea0c4c02d87d6cbfed8c/tools/utils.py#L242

However, we empirically found that using this step worsens the algorithm's performance in the relatively small-sized (e.g., 8x8 and 32x32) MIMO systems we considered in our work. This can be partly attributed to the fact that the theory of OAMP (including the orth step) is derived under the assumption of the large system limit. In realistic MIMO systems, this assumption may not be satisfied and that is why we use deep learning to enhance the algorithm.

For the second concern, our "neural network" is an unfolded one, which belongs to the concept of deep unfolding. This model-driven deep learning method is quite different from the traditional understanding of neural networks. Please refer to [1] and [2] for a further understanding of this methodology.

[1] V. Monga, Y. Li, and Y. C. Eldar, “Algorithm unrolling: Interpretable, efficient deep learning for signal and image processing,” IEEE Signal Process. Mag., vol. 38, no. 2, pp. 18–44, Mar. 2021.
[2] N. Shlezinger, J. Whang, Y. C. Eldar, and A. G. Dimakis, “Modelbased deep learning,” Proc. IEEE, vol. 111, no. 5, pp. 465–499, May 2023.

Best regards,
Xingyu

**Comment by jiadong5** [comment]
Hi Xingyu,

Thx for your reply, actually I have done some performance simulations on the small-size QPSK modulation in 64*64 situation, here are the results:
![image](https://github.com/STARainZ/CG-OAMP-NET/assets/83275455/efd6260f-729e-412c-b437-0042316394cd)
AMP_exact_variance is just oamp with no orth step as your code and updating the estimation with true variance, and I found that at least for 64 * 64 an orth step helps a lot.

For my second concern on NN structure, I read the two papers you have listed and fully understand unrolling an algorithm would lead to a Neural Network-like structure. Thx for providing me those papers, I totally agree that there is a neural network structure now.

Best Wishes,
Jade.

**Comment by STARainZ** [comment]
Hi Jade,

Thank you for reporting the issue. I have tried the 64x64 MIMO case with QPSK and found a similar result in BERs as your implementation, that is, OAMP with the orthogonal step has an advantage. Here is the result:
![OAMP_w_wo_orth](https://github.com/STARainZ/CG-OAMP-NET/assets/37243940/a6a047bc-3d25-49f2-bb99-c23d55b24099)

In this figure, I also show the 8x8 MIMO case with 16-QAM, where the OAMP Orth.'s performance falls behind (at least in our implementation), as I have commented above. 

Best regards,
Xingyu
